pageserver/src/walingest.rs

```diff
@@ -613,20 +613,18 @@ impl<'a> WalIngest<'a> {
                 forknum: FSM_FORKNUM,
             };
 
-            // FIXME: 'blkno' stored in the WAL record is the new size of the
-            // heap. The formula for calculating the new size of the FSM is
-            // pretty complicated (see FreeSpaceMapPrepareTruncateRel() in
-            // PostgreSQL), and we should also clear bits in the tail FSM block,
-            // and update the upper level FSM pages. None of that has been
-            // implemented. What we do instead, is always just truncate the FSM
-            // to zero blocks. That's bad for performance, but safe. (The FSM
-            // isn't needed for correctness, so we could also leave garbage in
-            // it. Seems more tidy to zap it away.)
-            if rec.blkno != 0 {
-                info!("Partial truncation of FSM is not supported");
+            let fsm_logical_page_no = rec.blkno / pg_constants::SLOTS_PER_FSM_PAGE;
+            let mut fsm_physical_page_no = fsm_logical_to_physical(fsm_logical_page_no);
+            if rec.blkno % pg_constants::SLOTS_PER_FSM_PAGE != 0 {
+                // Tail of last remaining FSM page has to be zeroed.
+                // We are not doing it here: there should be FPI wal-record for this page stored sometimes laters
+                fsm_physical_page_no += 1;
+            }
```

Comment:
Shouldn't we also update upper level FSM pages as removed comment said?

Reply:
We do not update pages at all at this point.
This is what I tried to say in the comment I have added instead of removed.
We are just updating relation size here (it is separate key-value entry in pageserver storage).

All FSM/VM pages are updated t compute nodes and when them are written (when smgrwrite is called for them), 
pagestore_smgr enforces wal-logging of FSM/VM pages so that their content is transferred to pageserver.